### PR TITLE
CI: add pytest for LAI

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

adding pytest for the full `lightning` package

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @tchaton @rohitgr7 @carmocca @akihironitta @borda

### CODE REVIEW

File: .actions/assistant.py [hunk 1]
```diff
@@ -163,6 +164,27 @@ def mirror_pkg2source(pypi_folder: str, src_folder: str) -> None:
                 shutil.rmtree(py_dir2, ignore_errors=True)
                 shutil.copytree(py_dir, py_dir2)
 
+    @staticmethod
+    def copy_replace_imports(
+        source_dir: str, source_import: str, target_import: str, target_dir: Optional[str] = None
```
Comment (otaj) [inline .actions/assistant.py:169]: Can we have a test for this function, please? I'm not 100% sure it covers all possible edge cases (i.e. recently introduced graveyard)

Comment (Borda) [reply .actions/assistant.py:169]: Agree, so run a test on our codebase?

Comment (otaj) [reply .actions/assistant.py:169]: I meant a pytest function, so that we can have it tested all the time in CI so that we know we didn't introduce any new pattern which would break this

Comment (Borda) [reply .actions/assistant.py:169]: I see... We shall add a test for most of the Assistant methods... :innocent:

File: requirements/app/base.txt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 lightning-cloud>=0.5.10
-packaging
+typing-extensions>=4.0.0, <4.3.1
```
Comment (Borda) [inline requirements/app/base.txt:2]: Seems like this is causing some problems. Any idea @otaj?
https://github.com/Lightning-AI/lightning/actions/runs/3241245486/jobs/5312932676#step:12:38

File: requirements/lite/base.txt [hunk 3]
```diff
@@ -3,7 +3,7 @@
 
 numpy>=1.17.2, <1.23.1
 torch>=1.9.*, <1.13.0
-fsspec[http]>=2021.06.1, <2022.6.0
+fsspec[http]>2021.06.0, <2022.6.0
```
Comment (Borda) [inline requirements/lite/base.txt:6]: this is a hack to the system as for oldest we replace `>=` by `==` and the the fixed version from PL and App was not compatible so I made this lower and `>` only but the meaning/range is identical

File: src/lightning_app/cli/app-template/tests/test_placeholdername_app.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
 import os
 from contextlib import redirect_stdout
 
-from lightning.app.testing.testing import application_testing, LightningTestApp
+from lightning_app.testing.testing import application_testing, LightningTestApp
```
Comment (Borda) [inline src/lightning_app/cli/app-template/tests/test_placeholdername_app.py:13]: Wondering how we were bel to test App as stand-alone when these imports shall be breaking it...

File: src/lightning_app/cli/pl-app-template/app.py [hunk 5]
```diff
@@ -4,10 +4,10 @@
 from core.components import TensorBoard, WeightsAndBiases
 from core.components.script_runner import ScriptRunner
 
-from lightning.app import LightningApp, LightningFlow
-from lightning.app.frontend import StaticWebFrontend
-from lightning.app.storage import Path
-from lightning.app.utilities.packaging.cloud_compute import CloudCompute
+from lightning_app import LightningApp, LightningFlow
```
Comment (Borda) [inline src/lightning_app/cli/pl-app-template/app.py:7]: Also, this will be replaced back when you'll be using `lightning` package, coming in #15105

File: requirements/app/base.txt (hunk 2 above)
Comment (otaj) [reply requirements/app/base.txt:2]: What? How?

Comment (otaj) [reply requirements/app/base.txt:2]: I mean, I see that we are failing on this step, but wtf?

File: src/lightning_app/frontend/panel/panel_frontend.py [hunk 6]
```diff
@@ -50,7 +50,7 @@ class PanelFrontend(Frontend):
     .. code-block:: python
 
         import lightning as L
-        from lightning.app.frontend.panel import PanelFrontend
+        from lightning_app.frontend.panel import PanelFrontend
```
Comment (awaelchli) [inline src/lightning_app/frontend/panel/panel_frontend.py:53]: Why are these being rewritten? In all documentation and examples, we need to show the lightning import. Here with the frontend, the problem is that frontend is not a top-level importable module. Either frontend has to be added in `__init__.py` or change the import to import `lightning.app.frontend` as frontend or similar. 


See also #15096

**Review by awaelchli — DISMISSED** [review]

File: src/lightning_app/frontend/panel/panel_frontend.py (hunk 6 above)
Comment (justusschock) [reply src/lightning_app/frontend/panel/panel_frontend.py:53]: @awaelchli this way lightning_app could theoretically be installed as a standalone package. This is required for the import-converter to work properly and to be able to test the packages independently. In the end the import-converter makes sure that this will be `lightning.app`

Comment (awaelchli) [reply src/lightning_app/frontend/panel/panel_frontend.py:53]: Does the import converter somehow modify the docstrings before docs are generated?
What about the examples that people see in our GitHub examples/ directory?

File: requirements/app/base.txt (hunk 2 above)
Comment (otaj) [reply requirements/app/base.txt]: Once again, we *NEED* to have tests for the magic in `assistant.py` and `setup_tools.py`. This error is caused by this: https://github.com/Lightning-AI/lightning/actions/runs/3241541479/jobs/5313622020#step:6:12

File: .azure/gpu-tests-lite.yml [hunk 7]
```diff
@@ -36,6 +36,7 @@ pr:
       - "tests/tests_lite/run_standalone_tests.sh"  # a symlink to the one above
 
 jobs:
+  # TODO: parametrize for using LAI mono-package
```
Comment (carmocca) [inline .azure/gpu-tests-lite.yml:39]: What does this TODO mean?

File: .github/workflows/ci-lai-test-full.yml [hunk 8]
```diff
@@ -0,0 +1,235 @@
+name: Test LAI full
```
Comment (carmocca) [inline .github/workflows/ci-lai-test-full.yml:1]: What's the benefit of having a monolithic workflow to run everything?

This makes things unnecessarily complicated, the filters need to account for everything, there's skip and conditions everywhere

Comment (otaj) [reply .github/workflows/ci-lai-test-full.yml:1]: I think since we're doing mangling of paths in creating LAI package, we need to be sure, everything works as expected.

File: src/lightning_app/frontend/panel/panel_frontend.py (hunk 6 above)
Comment (Borda) [reply src/lightning_app/frontend/panel/panel_frontend.py:53]: Yes, but if you install `lightning-app`, not `lightning` it will crash, in such case we deny users to install `lightning-app`

File: .azure/gpu-tests-lite.yml (hunk 7 above)
Comment (Borda) [reply .azure/gpu-tests-lite.yml:39]: will be added in separate PR

File: .github/workflows/ci-lai-test-full.yml (hunk 8 above)
Comment (Borda) [reply .github/workflows/ci-lai-test-full.yml:1]: @justusschock mentioned that we somewhere override sys path so we need to be sure that it will be working even from the created mono-package

File: src/lightning_app/frontend/panel/panel_frontend.py (hunk 6 above)
Comment (Borda) [reply src/lightning_app/frontend/panel/panel_frontend.py:53]: and in the mono-package it will be back as `lightning.app`

Comment (Borda) [reply src/lightning_app/frontend/panel/panel_frontend.py:53]: see #15105

Comment (awaelchli) [reply src/lightning_app/frontend/panel/panel_frontend.py:53]: I thought that you always get the lightning meta package. If you install pytorch-lightning you get lightning.pytorch but not lightning.app. is this not what we are going for?

File: .azure/gpu-tests-lite.yml (hunk 7 above)
Comment (carmocca) [reply .azure/gpu-tests-lite.yml:39]: Sure but what does it mean?

File: .github/workflows/ci-lai-test-full.yml (hunk 8 above)
Comment (carmocca) [reply .github/workflows/ci-lai-test-full.yml:1]: I don't understand how that relates to merging all testing workflows

Comment (justusschock) [reply .github/workflows/ci-lai-test-full.yml:1]: @carmocca it does not, we will split them after we are successfully running them!

**Review by awaelchli — APPROVED** [review]

File: src/lightning_app/frontend/panel/panel_frontend.py (hunk 6 above)
Comment (Borda) [reply src/lightning_app/frontend/panel/panel_frontend.py]: not really; these are the cases:
- install `pytorch-lightning` you get `pytorch_lightning` only (now from master you get also `lightning_lite`)
- install `lightning-app` you get `lightning_app` only
- install `lightning` you get all there:
  * `pytorch_lightning` and `lightning.pytorch`
  * `lightning_app` and `lightning.app`

**Review by justusschock — CHANGES_REQUESTED** [review]
Merge Blocker. @Borda and I synced and came up to a mixed solution of #15209 and this one which we both think to be even better
